builtin/providers/aws/resource_aws_route53_record.go

```diff
@@ -287,55 +310,18 @@ func resourceAwsRoute53RecordRead(d *schema.ResourceData, meta interface{}) erro
 		if record.SetIdentifier != nil && *record.SetIdentifier != d.Get("set_identifier") {
 			continue
 		}
-
-		found = true
-
-		err := d.Set("records", flattenResourceRecords(record.ResourceRecords))
-		if err != nil {
-			return fmt.Errorf("[DEBUG] Error setting records for: %s, error: %#v", en, err)
-		}
-
-		d.Set("ttl", record.TTL)
-		// Only set the weight if it's non-nil, otherwise we end up with a 0 weight
-		// which has actual contextual meaning with Route 53 records
-		//   See http://docs.aws.amazon.com/fr_fr/Route53/latest/APIReference/API_ChangeResourceRecordSets_Examples.html
-		if record.Weight != nil {
-			d.Set("weight", record.Weight)
-		}
-		d.Set("set_identifier", record.SetIdentifier)
-		d.Set("failover", record.Failover)
-		d.Set("health_check_id", record.HealthCheckId)
-
-		break
+		return record, nil
 	}
-
-	if !found {
-		log.Printf("[DEBUG] No matching record found for: %s, removing from state file", en)
-		d.SetId("")
-	}
-
-	return nil
+	return nil, fmt.Errorf("nothing found")
 }
 
 func resourceAwsRoute53RecordDelete(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).r53conn
-
-	zone := cleanZoneID(d.Get("zone_id").(string))
-	log.Printf("[DEBUG] Deleting resource records for zone: %s, name: %s",
-		zone, d.Get("name").(string))
-	var err error
-	zoneRecord, err := conn.GetHostedZone(&route53.GetHostedZoneInput{Id: aws.String(zone)})
-	if err != nil {
-		if r53err, ok := err.(awserr.Error); ok && r53err.Code() == "NoSuchHostedZone" {
-			log.Printf("[DEBUG] No matching Route 53 Record found for: %s, removing from state file", d.Id())
-			d.SetId("")
-			return nil
-		}
-		return err
-	}
 	// Get the records
-	rec, err := resourceAwsRoute53RecordBuildSet(d, *zoneRecord.HostedZone.Name)
+	rec, err := findRecord(d, meta)
 	if err != nil {
+		log.Printf("[DEBUG] No matching record found for: %s, removing from state file", d.Id())
+		d.SetId("")
 		return err
```

Comment:
Looks like any kind of error here is treated the same. Up in `findRecord` there's code to recognize `NoSuchHostedZone` specifically, but then it's wrapped back up in a regular `error` return value.

I believe this would mean that, for example, if ran `terraform plan` w/o network connectivity, I could lose all my R53 records from the state.

Maybe we can change the `"NoSuchHostedZone"` case to just return `nil`, `nil` and switch this up to something like:

```go
if err != nil {
  return err
}
if rec == nil {
  log.Printf("[WARN] No matching record found...")
  d.SetId("")
}
```

What do you think?